components/service-operator/internal/aws/cloudformation/client.go

```diff
@@ -0,0 +1,477 @@
+		}
+		if state.StackName != nil {
+			s.AWS.Name = *state.StackName
+		}
+		if state.StackStatus != nil {
+			s.AWS.Status = *state.StackStatus
+		}
+		if state.StackStatusReason != nil {
+			s.AWS.Reason = *state.StackStatusReason
+		}
+	}
+	// add any event details
+	if events != nil && len(events) > 0 {
+		s.AWS.Events = []object.AWSEvent{}
+		for _, event := range events {
+			reason := "-"
+			if event.ResourceStatusReason != nil {
+				reason = *event.ResourceStatusReason
+			}
+			s.AWS.Events = append(s.AWS.Events, object.AWSEvent{
+				Status: *event.ResourceStatus,
+				Reason: reason,
+				Time:   &metav1.Time{Time: *event.Timestamp},
+			})
+		}
+	}
+	// update generic state
+	switch s.AWS.Status {
+	case DeleteFailed, CreateFailed, RollbackFailed, UpdateRollbackFailed, RollbackComplete, UpdateRollbackComplete:
+		s.State = object.ErrorState
+	case DeleteInProgress, DeleteComplete:
+		s.State = object.DeletingState
+	case CreateComplete, UpdateComplete:
+		s.State = object.ReadyState
+	default:
+		s.State = object.ReconcilingState
+	}
+	// if object implements whitelisting of output keys, then update info
+	if w, ok := stack.(StackOutputWhitelister); ok {
+		if s.AWS.Info == nil {
+			s.AWS.Info = map[string]string{}
+		}
+		outputs, _ := r.Outputs(ctx, stack)
+		for _, whitelistedKey := range w.GetStackOutputWhitelist() {
+			s.AWS.Info[whitelistedKey] = outputs[whitelistedKey]
+		}
+	}
+	stack.SetStatus(s)
+}
+
+func (r *Client) events(ctx context.Context, stack Stack) ([]*StateEvent, error) {
+	eventsOutput, err := r.Client.DescribeStackEventsWithContext(ctx, &DescribeStackEventsInput{
+		StackName: aws.String(stack.GetStackName()),
+	})
+	if err != nil {
+		return nil, err
+	}
+	if eventsOutput == nil {
+		return []*StateEvent{}, nil
+	}
+	return eventsOutput.StackEvents, nil
+}
+
+// Exists checks if the stack has been provisioned
+func (r *Client) exists(ctx context.Context, stack Stack) (bool, error) {
+	_, err := r.get(ctx, stack)
+	if err == ErrStackNotFound {
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (r *Client) waitUntilCompleteState(ctx context.Context, stack Stack) (*State, error) {
+	return r.waitUntilState(ctx, stack, []string{
+		CreateComplete,
+		UpdateComplete,
+		UpdateRollbackComplete,
+		RollbackComplete,
+	})
+}
+
+func (r *Client) waitUntilDestroyedState(ctx context.Context, stack Stack) (*State, error) {
+	return r.waitUntilState(ctx, stack, []string{
+		DeleteComplete,
+	})
+}
+
+func (r *Client) waitUntilState(ctx context.Context, stack Stack, desiredStates []string) (*State, error) {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, context.DeadlineExceeded
+		default:
+			state, err := r.get(ctx, stack)
+			if IsNotFoundError(err) && in(DeleteComplete, desiredStates) {
+				// If we are waiting for DeleteComplete state and the
+				// stack has gone missing, consider this DeleteComplete
+				return &State{}, nil
```

Comment:
Shouldn't this be a string like DELETE_COMPLETE, or DeleteComplete?

Reply:
This function's job is to wait for a state and then return the entire cloudformation stack data object (State) ... When the stack has been deleted we don't have a stack data object, but it's also not an error... The best it can do is return a blank data object which is slightly more representative than nil

It is a bit weird ... But the DELETE_COMPLETE string will only show for about a second before vanishing so can't be trusted.

The cloudformation.State type/alias is a bit of a confusing name, but it was slightly less confusing than two different "Stack" types

Reply:
Ohhh... It's actually an empty stack. Let's see if we can find a better alternative to `type State = cloudformation.Stack` on Monday.

Reply:
We'll return to this in a later PR.